### ISSUE TITLE
Oneof/AnyOf/AllOf applied to the properties of an object

### ISSUE THREAD

**Issue by CMCDragonkai** [justinrainbow/json-schema#393]
I have tried these schemas based on these answers:

* http://stackoverflow.com/questions/31839578/how-to-define-a-json-schema-that-requires-at-least-one-of-many-properties
* http://stackoverflow.com/questions/25014650/json-schema-example-for-oneof-objects

But I can't get them to work:

```php
<?php

use JsonSchema\Validator;
use JsonSchema\Constraints\Constraint;
use JsonSchema\Exception\ValidationException;

$validator = new Validator;

$data = file_get_contents('./test.json');

$data = json_decode($data, true);

$validator->validate(
    $data,
    json_decode(json_encode([
        'type' => 'object',
        'properties' => [
            'response' => [
                'type' => 'object',
                'oneOf' => [
                    ['required' => ['error']],
                    ['required' => ['result']]
                ],
                'properties' => [
                    'error' => [
                        'type' => 'object'
                    ],
                    'result' => [
                        'type' => 'object'
                    ]
                ]
            ]
        ],
        'required' => ['response']
    ])),
    Constraint::CHECK_MODE_TYPE_CAST | Constraint::CHECK_MODE_EXCEPTIONS
);

var_dump($validator->isValid());
// JsonSchema\Exception\ValidationException with message 'Error validating /response/error: The property error is required'
```

```php
<?php

use JsonSchema\Validator;
use JsonSchema\Constraints\Constraint;
use JsonSchema\Exception\ValidationException;

$validator = new Validator;

$data = file_get_contents('./test.json');

$data = json_decode($data, true);

$validator->validate(
    $data,
    json_decode(json_encode([
        'type' => 'object',
        'properties' => [
            'response' => [
                'type' => 'object',
                'oneOf' => [
                    [
                        'properties' => [
                            'error' => [
                                'type' => 'object'
                            ],
                        ]
                    ],
                    [
                        'properties' => [
                            'result' => [
                                'type' => 'object'
                            ],
                        ],
                    ]
                ]
            ]
        ],
        'required' => ['response']
    ])),
    Constraint::CHECK_MODE_TYPE_CAST | Constraint::CHECK_MODE_EXCEPTIONS
);

var_dump($validator->isValid());
// JsonSchema\Exception\ValidationException with message 'Error validating /response: Failed to match exactly one schema'
```

The data is like:

```
{
    "response": {
        "result": {}
    }
}
```

```
{
    "response": {
        "error": {}
    }
}
```

**Comment by erayd** [comment]
Investigating this.

For anyone else following, cleaned up schemas from the original post:
```JSON
{
    "type": "object",
    "properties": {
        "response": {
            "type": "object",
            "oneOf": [
                {
                    "required": [
                        "error"
                    ]
                },
                {
                    "required": [
                        "result"
                    ]
                }
            ],
            "properties": {
                "error": {
                    "type": "object"
                },
                "result": {
                    "type": "object"
                }
            }
        }
    },
    "required": [
        "response"
    ]
}
```
```JSON
{
    "type": "object",
    "properties": {
        "response": {
            "type": "object",
            "oneOf": [
                {
                    "properties": {
                        "error": {
                            "type": "object"
                        }
                    }
                },
                {
                    "properties": {
                        "result": {
                            "type": "object"
                        }
                    }
                }
            ]
        }
    },
    "required": [
        "response"
    ]
}
```

**Comment by CMCDragonkai** [comment]
This one isn't working either: 

```php
<?php

use JsonSchema\Validator;
use JsonSchema\Constraints\Constraint;
use JsonSchema\Exception\ValidationException;

$validator = new Validator;

$data = file_get_contents('./test2.json');

$data = json_decode($data, true);

$validator->validate(
    $data,
    json_decode(json_encode([
        'type' => 'object',
        'properties' => [
            'response' => [
                'oneOf' => [
                    ['type' => 'object', 'properties' => [
                        'error' => [
                            'type' => 'object'
                        ],
                        'required' => ['error']
                    ]],
                    ['type' => 'object', 'properties' => [
                        'result' => [
                            'type' => 'object'
                        ],
                        'required' => ['result']
                    ]]
                ]
            ]
        ],
        'required' => ['response']
    ])),
    Constraint::CHECK_MODE_TYPE_CAST | Constraint::CHECK_MODE_EXCEPTIONS
);

var_dump($validator->isValid());
```

**Comment by erayd** [comment]
@CMCDragonkai Your first example works fine:
```PHP
<?php

require('vendor/autoload.php');

use JsonSchema\Validator;

$v = new Validator();

$schema_a = json_decode(file_get_contents('schema-a.json'));
$data_a=json_decode('{"response":{"result":{}}}');

$v->validate($data_a, $schema_a);
assert($v->isValid()); //true
```

**Comment by CMCDragonkai** [comment]
Maybe something's going wrong with the json encode and decode that I'm using to avoid typing out `(object)`.

**Comment by erayd** [comment]
@CMCDragonkai Your second example works fine too - it fails validation, as it should, because your input data matches both schemas in `oneOf`.

Your third example has an invalid schema - properties must be objects, but you have defined one named "required" as an array.

Could you please explain exactly what your issue is? Because from where I'm sitting, this doesn't look like a problem with the library, or with `json_encode` / `json_decode`. This looks like user error on your part.

You may wish to check out http://www.jsonschemavalidator.net/ to assist with writing a valid schema that fulfils your requirements.

**Comment by CMCDragonkai** [comment]
Try this:

```php
<?php

use JsonSchema\Validator;
use JsonSchema\Constraints\Constraint;
use JsonSchema\Exception\ValidationException;

$validator = new Validator;

$data = <<<'JSON'
{
    "response": {
        "result": {}
    }
}
JSON;

$data = json_decode($data, true);

$validator->validate(
    $data,
    (object)[
        'type' => 'object',
        'properties' => (object)[
            'response' => (object)[
                'type' => 'object',
                'oneOf' => [
                    (object)['required' => ['error']],
                    (object)['required' => ['result']]
                ],
                'properties' => (object)[
                    'error' => (object)[
                        'type' => 'object'
                    ],
                    'result' => (object)[
                        'type' => 'object'
                    ]
                ]
            ]
        ],
        'required' => ['response']
    ],
    Constraint::CHECK_MODE_TYPE_CAST | Constraint::CHECK_MODE_EXCEPTIONS
);

var_dump($validator->isValid());
```

**Comment by CMCDragonkai** [comment]
It's possible that because of `{}` gets decoded as a PHP array with the `true` option, and I'm looking for an object.

**Comment by CMCDragonkai** [comment]
I'm on 5.1 version as well, and my most recent example really doesn't work. Even if I remove the `true` and `Constraint::CHECK_MODE_TYPE_CAST`, it still throws an error.

**Comment by CMCDragonkai** [comment]
> Your third example has an invalid schema - properties must be objects, but you have defined one named "required" as an array.

My third example has `properties` as an object, not sure what you mean that I've 'defined one named "required" as an array'? Because I haven't done that. This is the internal-memory output from the third schema in PHP:

```
{#238
     +"type": "object",
     +"properties": {#246
       +"response": {#245
         +"oneOf": [
           {#241
             +"type": "object",
             +"properties": {#242
               +"error": {#243
                 +"type": "object",
               },
               +"required": [
                 "error",
               ],
             },
           },
           {#180
             +"type": "object",
             +"properties": {#244
               +"result": {#240
                 +"type": "object",
               },
               +"required": [
                 "result",
               ],
             },
           },
         ],
       },
     },
     +"required": [
       "response",
     ],
   }
```

None of the properties there are typed as `arrays`. They are all `objects` because of their object id.

**Comment by CMCDragonkai** [comment]
I found the problem. It's because of `Constraint::CHECK_MODE_EXCEPTIONS`. If you add it in, it fails on the first error. That error is that it fails one of the of `oneOf`. If you remove `CHECK_MODE_EXCEPTIONS`, it succeeds at the very end.

**Comment by CMCDragonkai** [comment]
Compare:

```php
<?php

use JsonSchema\Validator;
use JsonSchema\Constraints\Constraint;
use JsonSchema\Exception\ValidationException;

$validator = new Validator;

$data = <<<'JSON'
{
    "response": {
        "result": {
        }
    }
}
JSON;

$schema = <<<'JSON'
{
    "type": "object",
    "properties": {
        "response": {
            "type": "object",
            "oneOf": [
                {
                    "required": [
                        "error"
                    ]
                },
                {
                    "required": [
                        "result"
                    ]
                }
            ],
            "properties": {
                "error": {
                    "type": "object"
                },
                "result": {
                    "type": "object"
                }
            }
        }
    },
    "required": [
        "response"
    ]
}
JSON;

$data = json_decode($data);
$schema = json_decode($schema);

var_dump($data);
var_dump($schema);

$validator->validate(
    $data,
    $schema
);

var_dump($validator->isValid());
```

```
<?php

use JsonSchema\Validator;
use JsonSchema\Constraints\Constraint;
use JsonSchema\Exception\ValidationException;

$validator = new Validator;

$data = <<<'JSON'
{
    "response": {
        "result": {
        }
    }
}
JSON;

$schema = <<<'JSON'
{
    "type": "object",
    "properties": {
        "response": {
            "type": "object",
            "oneOf": [
                {
                    "required": [
                        "error"
                    ]
                },
                {
                    "required": [
                        "result"
                    ]
                }
            ],
            "properties": {
                "error": {
                    "type": "object"
                },
                "result": {
                    "type": "object"
                }
            }
        }
    },
    "required": [
        "response"
    ]
}
JSON;

$data = json_decode($data);
$schema = json_decode($schema);

var_dump($data);
var_dump($schema);

$validator->validate(
    $data,
    $schema,
    Constraint::CHECK_MODE_EXCEPTIONS
);

var_dump($validator->isValid());
```

**Comment by shmax** [comment]
try:

```PHP
$data = json_decode($data, false);
```

**Comment by CMCDragonkai** [comment]
@shmax Please see my most recent example. It shows that the problem is with exception constraint, not about json decoding.

**Comment by erayd** [comment]
Thanks - I can confirm there is a problem with `CHECK_MODE_EXCEPTIONS`; I will fix this.

Regarding the other issues you're having - you are definitely making mistakes with some of your schemas. I strongly recommend you take advantage of http://www.jsonschemavalidator.net/; it will help you ensure that your schemas are valid.

> My third example has properties as an object, not sure what you mean that I've 'defined one named "required" as an array'? Because I haven't done that.

Yes, you have done that. You have `properties->required->[]`, which is illegal.

**Comment by CMCDragonkai** [comment]
@erayd I'm sorry but I just have my Spacemacs telling me I do not have `required` as a property of `properties`. They are siblings, not parent-children relationship.

Thank you for the fix.

**Comment by erayd** [comment]
@CMCDragonkai 

Look at `properties>>response>>oneOf>>properties>>required` [here](https://github.com/justinrainbow/json-schema/issues/393).

I have now pointed this out three times. Look at your code.

I know there is a bug with `CHECK_MODE_EXCEPTIONS`, however your insistence that there is no user-error here is annoying, and makes me less inclined to pay attention when you submit new issues. Please check your input before claiming it's perfect, especially when someone has already told you multiple times it's not.

**Comment by CMCDragonkai** [comment]
@erayd I see it. I thought you were referring to the first "required" property, not the inner set.

**Comment by erayd** [comment]
The `CHECK_MODE_EXCEPTIONS` bug has been fixed in #394. Thanks for bringing it to my attention :-).
